File: plugins/grafana-notion-datasource/pkg/plugin/filter.go
package plugin

import (
	"strconv"
	"strings"
)

// FilterNode is a node in the structured filter tree sent from the query editor.
// A node is either a condition (kind="condition") or a group (kind="group").
//
// Unlike a SQL-style where clause, Notion filters are JSON objects. Each
// condition compiles to {"property": <name>, <typeKey>: {<op>: <value>}} and
// groups compile to {"and": [...]} / {"or": [...]}. The condition therefore
// needs the property's type category (text/number/checkbox/date/select/...) to
// pick the correct type key and to coerce the value.
type FilterNode struct {
	Kind string `json:"kind"`

	// Condition fields.
	Field string `json:"field,omitempty"`
	// Category is the logical property category (text, number, checkbox, date,
	// select, multi_select, people, files). It selects the Notion filter type key.
	Category string `json:"category,omitempty"`
	Op       string `json:"op,omitempty"`
	Value    string `json:"value,omitempty"`

	// Group fields.
	Connector string       `json:"connector,omitempty"`
	Children  []FilterNode `json:"children,omitempty"`
}

// typeKeyForCategory maps a logical property category to the Notion filter
// object key. Notion nests the operator under a key named after the property
// type, e.g. {"property":"Name","rich_text":{"equals":"x"}}.
func typeKeyForCategory(category string) string {
	switch category {
	case "number":
		return "number"
	case "checkbox":
		return "checkbox"
	case "date":
		return "date"
	case "select":
		return "select"
	case "status":
		return "status"
	case "multi_select":
		return "multi_select"
	case "people":
		return "people"
	case "files":
		return "files"
	case "text":
		fallthrough
	default:
		return "rich_text"
	}
}

// operatorArity classifies how many values an operator consumes:
//
//	"none"   – unary (no value), e.g. is_empty/is_not_empty
//	"single" – a single value (string/number/bool/date)
//	"list"   – comma-separated tokens compiled into a nested or-group of
//	           single-value conditions (Notion has no native "in" operator)
func operatorArity(op string) string {
	switch op {
	case "is_empty", "is_not_empty":
		return "none"
	case "in", "not_in":
		return "list"
	default:
		return "single"
	}
}

// BuildFilter converts a structured filter tree (the root group) into a Notion
// filter object suitable for the database query body. It returns nil when there
// are no valid conditions, so the caller can omit the filter entirely.
func BuildFilter(root *FilterNode) map[string]any {
	if root == nil {
		return nil
	}
	return compileGroup(root)
}

func compileNode(node FilterNode) map[string]any {
	if node.Kind == "group" {
		return compileGroup(&node)
	}
	return compileCondition(node)
}

func compileGroup(group *FilterNode) map[string]any {
	parts := make([]map[string]any, 0, len(group.Children))
	for _, child := range group.Children {
		if c := compileNode(child); c != nil {
			parts = append(parts, c)
		}
	}
	if len(parts) == 0 {
		return nil
	}
	if len(parts) == 1 {
		return parts[0]
	}
	connector := "and"
	if strings.EqualFold(group.Connector, "or") {
		connector = "or"
	}
	// Notion expects the array under the connector key. Each element may itself
	// be a condition or another and/or group.
	arr := make([]any, len(parts))
	for i, p := range parts {
		arr[i] = p
	}
	return map[string]any{connector: arr}
}

func compileCondition(c FilterNode) map[string]any {
	field := strings.TrimSpace(c.Field)
	if field == "" {
		return nil
	}
	op := strings.TrimSpace(c.Op)
	if op == "" {
		op = defaultOpForCategory(c.Category)
	}
	typeKey := typeKeyForCategory(c.Category)

	switch operatorArity(op) {
	case "none":
		// Unary operators are encoded as {<op>: true} under the type key.
		return condition(field, typeKey, map[string]any{op: true})
	case "list":
		tokens := splitTokens(c.Value)
		if len(tokens) == 0 {
			return nil
		}
		// Notion has no list membership operator, so expand into a group of
		// single-value conditions. "in" -> OR of equals; "not_in" -> AND of
		// not-equals.
		singleOp, connector := "equals", "or"
		if op == "not_in" {
			singleOp, connector = "does_not_equal", "and"
		}
		parts := make([]any, 0, len(tokens))
		for _, t := range tokens {
			parts = append(parts, condition(field, typeKey, map[string]any{singleOp: coerceValue(c.Category, t)}))
		}
		if len(parts) == 1 {
			return parts[0].(map[string]any)
		}
		return map[string]any{connector: parts}
	default: // single
		v := strings.TrimSpace(c.Value)
		if v == "" {
			return nil
		}
		return condition(field, typeKey, map[string]any{op: coerceValue(c.Category, v)})
	}
}

// condition builds a single Notion property filter object.
func condition(property, typeKey string, body map[string]any) map[string]any {
	return map[string]any{
		"property": property,
		typeKey:    body,
	}
}

// defaultOpForCategory returns a sensible default operator when none is given.
func defaultOpForCategory(category string) string {
	switch category {
	case "checkbox":
		return "equals"
	case "number", "date":
		return "equals"
	case "select", "status":
		return "equals"
	case "multi_select", "people":
		return "contains"
	default: // text
		return "equals"
	}
}

// coerceValue converts the string value from the editor into the JSON type
// Notion expects for the given category. Numbers become float64, checkbox
// becomes bool; everything else stays a string.
func coerceValue(category, value string) any {
	switch category {
	case "number":
		if f, err := strconv.ParseFloat(strings.TrimSpace(value), 64); err == nil {
			return f
		}
		return value
	case "checkbox":
		return parseBool(value)
	default:
		return value
	}
}

func parseBool(value string) bool {
	switch strings.TrimSpace(strings.ToLower(value)) {
	case "1", "true", "yes", "checked":
		return true
	default:
		return false
	}
}

func splitTokens(value string) []string {
	tokens := make([]string, 0)
	for _, t := range strings.Split(value, ",") {
		t = strings.TrimSpace(t)
		if t != "" {
			tokens = append(tokens, t)
		}
	}
	return tokens
}

File: plugins/datasources/yesoreyeram-notion-datasource/src/filter.ts
import { SelectableValue } from '@grafana/data';

// ---------------------------------------------------------------------------
// Filter model
// ---------------------------------------------------------------------------

export type LogicalConnector = 'and' | 'or';

export interface FilterCondition {
  kind: 'condition';
  /** Notion property name. */
  field: string;
  /** Logical property category used by the backend to pick the Notion filter type key. */
  category?: FieldCategory;
  /** Comparison operator (Notion operator name; see OPERATORS). */
  op: string;
  /** Comparison value. Ignored for unary operators; comma-separated for lists. */
  value?: string;
}

export interface FilterGroup {
  kind: 'group';
  /** How the direct children are combined. */
  connector: LogicalConnector;
  children: FilterNode[];
}

export type FilterNode = FilterCondition | FilterGroup;

export function newCondition(): FilterCondition {
  return { kind: 'condition', field: '', category: 'text', op: 'equals', value: '' };
}

export function newGroup(): FilterGroup {
  return { kind: 'group', connector: 'and', children: [] };
}

export function emptyRootGroup(): FilterGroup {
  return newGroup();
}

// ---------------------------------------------------------------------------
// Operator catalog (field-type aware)
// ---------------------------------------------------------------------------

export interface OperatorDef {
  value: string;
  label: string;
  /**
   * 'none'  = unary (no value)
   * 'single'= one value
   * 'list'  = comma-separated list (expanded into an or/and group server-side)
   */
  arity: 'none' | 'single' | 'list';
}

// Operator groups reused across field categories. The `value` is the exact
// Notion API operator name.
const EQUALITY: OperatorDef[] = [
  { value: 'equals', label: '=', arity: 'single' },
  { value: 'does_not_equal', label: '!=', arity: 'single' },
];

const NULLABLE: OperatorDef[] = [
  { value: 'is_empty', label: 'is empty', arity: 'none' },
  { value: 'is_not_empty', label: 'is not empty', arity: 'none' },
];

const NUMBER_COMPARISON: OperatorDef[] = [
  { value: 'greater_than', label: '>', arity: 'single' },
  { value: 'greater_than_or_equal_to', label: '>=', arity: 'single' },
  { value: 'less_than', label: '<', arity: 'single' },
  { value: 'less_than_or_equal_to', label: '<=', arity: 'single' },
];

const DATE_COMPARISON: OperatorDef[] = [
  { value: 'before', label: 'is before', arity: 'single' },
  { value: 'after', label: 'is after', arity: 'single' },
  { value: 'on_or_before', label: 'is on or before', arity: 'single' },
  { value: 'on_or_after', label: 'is on or after', arity: 'single' },
];

const TEXT_MATCH: OperatorDef[] = [
  { value: 'contains', label: 'contains', arity: 'single' },
  { value: 'does_not_contain', label: 'does not contain', arity: 'single' },
  { value: 'starts_with', label: 'starts with', arity: 'single' },
  { value: 'ends_with', label: 'ends with', arity: 'single' },
];

const SELECT_LIST: OperatorDef[] = [
  { value: 'in', label: 'is any of', arity: 'list' },
  { value: 'not_in', label: 'is none of', arity: 'list' },
];

const MULTI_OPS: OperatorDef[] = [
  { value: 'contains', label: 'contains', arity: 'single' },
  { value: 'does_not_contain', label: 'does not contain', arity: 'single' },
  { value: 'in', label: 'has any of', arity: 'list' },
  { value: 'not_in', label: 'has none of', arity: 'list' },
];

const BOOL: OperatorDef[] = [{ value: 'equals', label: '=', arity: 'single' }];

/** Logical categories a Notion property type can map to for operator selection. */
export type FieldCategory = 'text' | 'number' | 'checkbox' | 'date' | 'select' | 'status' | 'multi_select' | 'people' | 'files';

// Map raw Notion property types to a logical category.
const TYPE_CATEGORY: Record<string, FieldCategory> = {
  // text
  title: 'text',
  rich_text: 'text',
  email: 'text',
  phone_number: 'text',
  url: 'text',
  formula: 'text',
  rollup: 'text',
  relation: 'text',
  // number
  number: 'number',
  unique_id: 'number',
  // checkbox
  checkbox: 'checkbox',
  // date / time
  date: 'date',
  created_time: 'date',
  last_edited_time: 'date',
  // single select
  select: 'select',
  status: 'status',
  // multi / arrays
  multi_select: 'multi_select',
  people: 'people',
  created_by: 'people',
  last_edited_by: 'people',
  files: 'files',
};

export function categoryForType(type?: string): FieldCategory {
  if (!type) {
    return 'text';
  }
  return TYPE_CATEGORY[type] ?? 'text';
}

/** Returns the operator definitions valid for a given Notion property type. */
export function operatorsForType(type?: string): OperatorDef[] {
  switch (categoryForType(type)) {
    case 'number':
      return [...EQUALITY, ...NUMBER_COMPARISON, ...NULLABLE];
    case 'checkbox':
      return [...BOOL];
    case 'date':
      return [...EQUALITY, ...DATE_COMPARISON, ...NULLABLE];
    case 'select':
    case 'status':
      return [...EQUALITY, ...SELECT_LIST, ...NULLABLE];
    case 'multi_select':
    case 'people':
      return [...MULTI_OPS, ...NULLABLE];
    case 'files':
      return [...NULLABLE];
    case 'text':
    default:
      return [...EQUALITY, ...TEXT_MATCH, ...NULLABLE];
  }
}

export function operatorOptions(type?: string): Array<SelectableValue<string>> {
  return operatorsForType(type).map((o) => ({ label: o.label, value: o.value }));
}

export function operatorDef(op: string, type?: string): OperatorDef | undefined {
  return operatorsForType(type).find((o) => o.value === op) ?? ALL_OPERATORS.find((o) => o.value === op);
}

const ALL_OPERATORS: OperatorDef[] = [
  ...EQUALITY,
  ...NULLABLE,
  ...NUMBER_COMPARISON,
  ...DATE_COMPARISON,
  ...TEXT_MATCH,
  ...SELECT_LIST,
  { value: 'contains', label: 'contains', arity: 'single' },
  { value: 'does_not_contain', label: 'does not contain', arity: 'single' },
];

export function operatorArity(op: string): 'none' | 'single' | 'list' {
  return ALL_OPERATORS.find((o) => o.value === op)?.arity ?? 'single';
}

// ---------------------------------------------------------------------------
// Template interpolation
//
// The Notion filter object is built on the backend from the structured tree.
// Because backend data sources interpolate the query on the frontend before it
// is sent, we interpolate each condition value here (arity aware), so the
// backend receives concrete values.
// ---------------------------------------------------------------------------

export type ValueInterpolator = (value: string, asList: boolean) => string;

export function interpolateFilterTree(root: FilterGroup, interpolate: ValueInterpolator): FilterGroup {
  const mapNode = (node: FilterNode): FilterNode => {
    if (node.kind === 'group') {
      return { ...node, children: node.children.map(mapNode) };
    }
    if (!node.value) {
      return node;
    }
    const asList = operatorArity(node.op) === 'list';
    return { ...node, value: interpolate(node.value, asList) };
  };
  return { ...root, children: root.children.map(mapNode) };
}

// ---------------------------------------------------------------------------
// JSON (de)serialization for persistence on the query
// ---------------------------------------------------------------------------

export function parseFilterTree(raw?: string): FilterGroup {
  if (!raw) {
    return emptyRootGroup();
  }
  try {
    const parsed = JSON.parse(raw);
    if (parsed && parsed.kind === 'group' && Array.isArray(parsed.children)) {
      return parsed as FilterGroup;
    }
  } catch {
    // ignore malformed persisted state
  }
  return emptyRootGroup();
}

export function stringifyFilterTree(root: FilterGroup): string {
  // Empty tree persists as empty string to keep queries clean.
  if (root.children.length === 0) {
    return '';
  }
  return JSON.stringify(root);
}
